### PR TITLE
Add `UnionToEnum` and `TupleToEnum` types and fix `UnionToTuple` output

### LINKED ISSUE #1084: Union to Enum
### Type description + examples

When making dynamic enum from string template union:

```typescript
const x = ['create','read']
const y = ['other','foo']
type SomeTemplate = `${x}_${y}` // <-- A string union
const theUnion = x.flatMap(a => y.map(b => `${a}_${b}`))
const theEnum = Object.fromEntries(theUnion.map(a => [a,a]))
```

If we use `Record<SomeTemplate, SomeTemplate>`, the type of every key is `SomeTemplate` instead of the literal. Now I don't know how many will use it, but at least I did so I know that the enum is really just an enum instead of some other type. However, an argument can be made that the `Record<K, K>` is indication enough that it is an enum.

### Type source

```typescript
type UnionToEnum<T extends string> = {
  [k in T]: k;
};
```

### Search existing types and issues first

- [x] I tried my best to look for it

### PR DESCRIPTION
**Pull request by benzaria**
Closes #1084 

Add `UnionToEnum<T, Numeric, Options>` utility for converting a union/tuple into an Enum-like object.

 - `string`, `number`, or `symbol` unions or tuples
 - **Numeric Enum** values starting from configurable **Index**, default: `1`

**Example:**
```ts
import type {UnionToEnum} from 'type-fest';

type E1 = UnionToEnum<'A' | 'B' | 'C'>;
//=> { A: 'A'; B: 'B'; C: 'C' }

type E2 = UnionToEnum<'X' | 'Y' | 'Z', {numeric: true}>;
//=> { X: 1; Y: 2; Z: 3 }

type E3 = UnionToEnum<['Play', 'Pause', 'Stop'], {numeric: true; startIndex: 3}>;
//=> { Play: 3; Pause: 4; Stop: 5 }

type E4 = UnionToEnum<['some_key', 'another_key']>;
//=> { 'some_key': 'some_key'; 'another_key': 'another_key' }

type E5 = UnionToEnum<never>;
//=> {}
```

### CODE REVIEW

**Comment by benzaria** [comment]
@som-sm @sindresorhus Should **CamelCase** be `true` by default ?

**Comment by sindresorhus** [comment]
This type is doing too much. It should not do case transformation. That would be a task for a separate type. It should purely transform union to enum.

**Comment by benzaria** [comment]
> This type is doing too much. It should not do case transformation. That would be a task for a separate type. It should purely transform union to enum.

fair enough i will remove case transformation.

**Comment by sindresorhus** [comment]
I don't think it should do startingIndex either. It's too limiting and could be done by a separate type too.

**Comment by benzaria** [comment]
> I don't think it should do startingIndex either. It's too limiting and could be done by a separate type too.

Should I keep the starting index `1` or `0` ?

**Comment by sindresorhus** [comment]
Hmm, didn't think of that. Maybe we do indeed need the `startIndex`.

**Comment by benzaria** [comment]
> Hmm, didn't think of that. Maybe we do indeed need the `startIndex`.

True, diff users may want diff index for there enum.

**Comment by benzaria** [comment]
@sindresorhus @som-sm Ready !

File: source/union-to-enum.d.ts [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+//    readFile: 'read_file',
+//    readFolder: 'read_folder',
+//    readLink: 'read_link',
+//    deleteFile: 'delete_file',
+//    deleteFolder: 'delete_folder',
+//    deleteLink: 'delete_link',
+// }
+```
+
+@author benzaria
```
Comment (sindresorhus) [inline source/union-to-enum.d.ts]: When we comment something, we expect you to apply it to all your pull requests so we don't have to repeat it.

File: source/union-to-enum.d.ts [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+@see UnionToTuple
+@category Object
+*/
+export type UnionToEnum<
+	Keys extends PropertyKey | readonly PropertyKey[],
+	Numeric extends boolean = false,
```
Comment (sindresorhus) [inline source/union-to-enum.d.ts]: This would be better as an option.

File: readme.md [hunk 3]
```diff
@@ -245,6 +245,7 @@ Click the type names for complete docs.
 - [`ReadonlyTuple`](source/readonly-tuple.d.ts) - Create a type that represents a read-only tuple of the given type and length.
 - [`TupleToUnion`](source/tuple-to-union.d.ts) - Convert a tuple/array into a union type of its elements.
 - [`UnionToTuple`](source/union-to-tuple.d.ts) - Convert a union type into an unordered tuple type of its elements.
+- [`UnionToEnum`](source/union-to-enum.d.ts) - Convert a union or tuple of property keys into an **Enum**-like object.
```
Comment (sindresorhus) [inline readme.md]: "enum-like" is a bit vague.

File: source/union-to-enum.d.ts [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+import type {BuildTuple} from './internal/tuple.d.ts';
+import type {Simplify} from './simplify.d.ts';
+import type {IsNever} from './is-never.d.ts';
+
+/**
+{@link UnionToEnum} Options.
+*/
+type UnionToEnumOptions = {
+	/**
+	The starting **Index** of Numeric enums.
```
Comment (sindresorhus) [inline source/union-to-enum.d.ts]:
```suggestion
	The first numeric value to assign when using numeric indices.
```

File: test-d/union-to-enum.ts [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+	writeFile: 'write_file',
+	writeFolder: 'write_folder',
+	writeLink: 'write_link',
+	readFile: 'read_file',
+	readFolder: 'read_folder',
+	readLink: 'read_link',
+	deleteFile: 'delete_file',
+	deleteFolder: 'delete_folder',
+	deleteLink: 'delete_link',
+} as const);
```
Comment (sindresorhus) [inline test-d/union-to-enum.ts:70]: Some additional tests (AI generated):

```js
// Empty array cases
expectType<UnionToEnum<[]>>({});
expectType<UnionToEnum<readonly []>>({});

// Edge cases for startIndex
expectType<UnionToEnum<['a', 'b'], true, {startIndex: 0}>>({a: 0, b: 1} as const);
expectType<UnionToEnum<['x'], true, {startIndex: -1}>>({} as const);
expectType<UnionToEnum<['test'], true, {startIndex: 100}>>({test: 100} as const);

// Numeric edge cases
expectType<UnionToEnum<0 | -1 | 42>>({0: 0, [-1]: -1, 42: 42} as const);
expectType<UnionToEnum<[0, -5, 999], true>>({0: 1, [-5]: 2, 999: 3} as const);

// Mixed types with numeric indexing
expectType<UnionToEnum<['str', 42, 'another'], true>>({str: 1, 42: 2, another: 3} as const);

// Single element cases
expectType<UnionToEnum<[0]>>({0: 0} as const);
expectType<UnionToEnum<[0], true, {startIndex: 5}>>({0: 5} as const);

// Verify that readonly behavior is consistent
expectType<UnionToEnum<['a']>>({a: 'a'} as const);
```

File: source/union-to-enum.d.ts [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+type E5 = UnionToEnum<never>;
+//=> {}
+```
+
+@example
+```
+import type {UnionToEnum, CamelCasedProperties} from 'type-fest';
+
+const verb = ['write', 'read', 'delete'] as const;
+const resrc = ['file', 'folder', 'link'] as const;
```
Comment (sindresorhus) [inline source/union-to-enum.d.ts]: Don't use abbreviations

File: source/union-to-enum.d.ts (hunk 1 above)
Comment (benzaria) [reply source/union-to-enum.d.ts]: What is wrong with me. have I a short memory or some 🤦‍♂️

File: source/union-to-enum.d.ts (hunk 2 above)
Comment (benzaria) [reply source/union-to-enum.d.ts]: IDK, that will make it harder to reach. Numeric isn't an option as startIndex is, Numeric will be used quite often because its more of a feature of this type, anyone can make a mapped obj from a union with ease. the Numeric transformation is what make it standout. I think we should keep it with the generic arguments.

File: readme.md (hunk 3 above)
Comment (benzaria) [reply readme.md]: I tried to be clear, maybe to much.

File: test-d/union-to-enum.ts (hunk 5 above)
Comment (benzaria) [reply test-d/union-to-enum.ts:70]: Sure !

Comment (benzaria) [reply test-d/union-to-enum.ts:70]: The expected value does evaluate, but an error get thrown. Should I prevent this error from showing or keep it in.
```ts
expectType<UnionToEnum<['x'], true, {startIndex: -1}>>({} as const); //=> {}
but throw; Error: Type instantiation is excessively deep and possibly infinite. 
```

File: source/union-to-enum.d.ts [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+/**
+{@link UnionToEnum} Options.
+*/
+type UnionToEnumOptions = {
+	/**
+	The first numeric value to assign when using numeric indices.
+
+	@default 1
+	*/
+	startIndex?: number;
```
Comment (som-sm) [inline source/union-to-enum.d.ts]: Add example.

File: source/union-to-enum.d.ts (hunk 2 above)
Comment (som-sm) [reply source/union-to-enum.d.ts]: Should be an option, `UnionToEnum<'X' | 'Y' | 'Z', {numeric: true}>` is more clearer than `UnionToEnum<'X' | 'Y' | 'Z', true>`. I'd prefer typing in a few extra characters for better readability.

File: test-d/union-to-enum.ts (hunk 5 above)
Comment (som-sm) [reply test-d/union-to-enum.ts:70]: `-ve` `startIndex` should also work, you can use `Sum` instead of `BuildTuple`.

**Review by som-sm — COMMENTED** [review]
What should be the behaviour in cases where `UnionToEnum` returns an index signature?

For example, currently invalid index signatures are returned in cases like these:
```ts
type T1 = UnionToEnum<string | number>;
// {readonly [x: string]: string; readonly [x: number]: number}

type T2 = UnionToEnum<[string, 'foo'], true>;
// {readonly [x: string]: 1; readonly foo: 2}
```

image<br>

Maybe we should skip all non-literals.

File: test-d/union-to-enum.ts (hunk 5 above)
Comment (benzaria) [reply test-d/union-to-enum.ts:70]: Ohh good improvement. [030ce52](https://github.com/sindresorhus/type-fest/pull/1171/commits/030ce52bd2a03be291eb889df79f4ae42517ab87)

**Comment by benzaria** [comment]
> What should be the behaviour in cases where `UnionToEnum` returns an index signature?

Should also skip `foo${string}` !

File: source/union-to-enum.d.ts (hunk 2 above)
Comment (benzaria) [reply source/union-to-enum.d.ts]: moved to Options in [030ce52](https://github.com/sindresorhus/type-fest/pull/1171/commits/030ce52bd2a03be291eb889df79f4ae42517ab87)

File: source/union-to-enum.d.ts (hunk 7 above)
Comment (benzaria) [reply source/union-to-enum.d.ts]: Done in [030ce52](https://github.com/sindresorhus/type-fest/pull/1171/commits/030ce52bd2a03be291eb889df79f4ae42517ab87)

File: source/union-to-enum.d.ts [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+*/
+type UnionToEnumOptions = {
+	/**
+	The first numeric value to assign when using numeric indices.
+
+	@default 1
+
+	@example
+	```
+	type E2 = UnionToEnum<['Play', 'Pause', 'Stop'], {numeric: true}>;
```
Comment (som-sm) [inline source/union-to-enum.d.ts]: Should start with `E1`.

File: source/union-to-enum.d.ts [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+};
+
+type DefaultUnionToEnumOptions = {
+	startIndex: 1;
+	numeric: false;
+};
+
+/**
+Converts a union or tuple of property keys (string, number, or symbol) into an **Enum**.
+
```
Comment (sindresorhus) [inline source/union-to-enum.d.ts:6]: It should explicitly say that union ordering is not guaranteed.

File: .gitignore [hunk 10]
```diff
@@ -1,2 +1,3 @@
 node_modules
 yarn.lock
+\!*
```
Comment (sindresorhus) [inline .gitignore]: Unrelated change

**Review by sindresorhus — APPROVED** [review]

File: source/union-to-enum.d.ts (hunk 9 above)
Comment (benzaria) [reply source/union-to-enum.d.ts:6]: i added it to numeric.

File: .gitignore (hunk 10 above)
Comment (benzaria) [reply .gitignore]: im so sorry it was a mistake, i use it to mask some private tests and experimental types. i just forgot about it!

**Comment by som-sm** [comment]
Apologies, haven't been able to re-read this. I'll give it a read sometime next week.

**Comment by benzaria** [comment]
@som-sm, Hey 👋 just following up on this, you mentioned you’d review it next week, but I haven’t seen any updates yet.
No rush if you’re busy, I just wanted to check whether there’s anything else needed from my side to move this forward.

File: test-d/union-to-enum.ts [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+
+// Edge cases for startIndex
+expectType<UnionToEnum<['x'], {numeric: true; startIndex: -1}>>({x: -1} as const);
+expectType<UnionToEnum<['x', 'y'], {numeric: true; startIndex: -100}>>({x: -100, y: -99} as const);
+expectType<UnionToEnum<['test'], {numeric: true; startIndex: 100}>>({test: 100} as const);
+
+// Numeric edge cases
+expectType<UnionToEnum<0 | -1 | 42>>({0: 0, [-1]: -1, 42: 42} as const);
+expectType<UnionToEnum<[0, -5, 999], {numeric: true}>>({0: 1, [-5]: 2, 999: 3} as const);
```
Comment (sindresorhus) [inline test-d/union-to-enum.ts]: Maybe a few more tests:

```ts
// Symbols with numeric indices
expectType<UnionToEnum<[typeof sym1, typeof sym2], {numeric: true}>>({[sym1]: 1, [sym2]: 2} as const);

// Readonly tuple input
expectType<UnionToEnum<readonly ['a', 'b', 'c']>>({a: 'a', b: 'b', c: 'c'} as const);
```

Comment (benzaria) [reply test-d/union-to-enum.ts]: Done

**Comment by benzaria** [comment]
@som-sm Could you please take some time to review this PR when you get a chance, so we can move it forward?

File: source/union-to-enum.d.ts [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+//    deleteFolder: 'delete_folder',
+//    deleteLink: 'delete_link',
+// }
+```
+
+@see UnionToTuple
+@category Object
+*/
+export type UnionToEnum<
+	Keys extends PropertyKey | readonly PropertyKey[],
```
Comment (som-sm) [inline source/union-to-enum.d.ts]: This allows `Keys` to be a combination of `PropertyKey` and `readonly PropertyKey[]`, like:

```ts
type Current = UnionToEnum<'A' | ['B', 'C']>;
//=> { readonly A: "A"; }

type Ideal = UnionToEnum<'A' | ['B', 'C']>;
//=> { readonly A: "A"; readonly B: "B"; readonly C: "C"; }
```

File: source/union-to-enum.d.ts (hunk 8 above)
Comment (benzaria) [reply source/union-to-enum.d.ts]: Done!

File: source/union-to-enum.d.ts (hunk 12 above)
Comment (benzaria) [reply source/union-to-enum.d.ts]: Great catch! However, I think it’s better to restrict using both approaches (union and tuple) at the same time rather than allowing it.

The reason is that the behavior becomes unpredictable or effectively useless when `numeric` is `true`, since adding a union element to a tuple input defeats the purpose of numeric indexing.

Done in [7efbf20](https://github.com/sindresorhus/type-fest/pull/1171/commits/7efbf2066a1510ef40755d3615ab63d132f4ab43)

**Comment by Emiyaaaaa** [comment]
This type name and parameter make me confused, maybe it can be split to `UnionToEnum` and `TupleToEnum`? @sindresorhus @som-sm what do you think?

**Comment by som-sm** [comment]
> This type name and parameter make me confused, maybe it can be split to `UnionToEnum` and `TupleToEnum`? @sindresorhus @som-sm what do you think?

@Emiyaaaaa Yeah, pretty valid point, if the type is `UnionToEnum`, it shouldn't accept tuples. I think it makes sense to split it to `UnionToEnum` and `TupleToEnum`.

**Comment by benzaria** [comment]
@som-sm I also have a suggestion, why not just call it `ToEnum`. It's straight forward and eliminates the need for 2 types.

**Comment by som-sm** [comment]
> @som-sm I also have a suggestion, why not just call it `ToEnum`. It's straight forward and eliminates the need for 2 types.

Umm...no, `ToEnum` is quite vague, I think it's fine to split it into two types.
